crypto/rsa/rsa_gen.c

```diff
@@ -31,19 +31,43 @@ static int rsa_builtin_keygen(RSA *rsa, int bits, BIGNUM *e_value,
  */
 int RSA_generate_key_ex(RSA *rsa, int bits, BIGNUM *e_value, BN_GENCB *cb)
 {
-    if (rsa->meth->rsa_keygen)
+    if (rsa->meth->rsa_keygen != NULL)
         return rsa->meth->rsa_keygen(rsa, bits, e_value, cb);
-    return rsa_builtin_keygen(rsa, bits, e_value, cb);
+
+    return rsa_builtin_keygen(rsa, bits, RSA_DEFAULT_PRIME_NUM, e_value, cb);
```

Comment:
Actually, I would call `RSA_generate_multi_prime_key` here.  That way, we can allow someone to implement `rsa->meth->rsa_multi_prime_keygen` only...

Reply:
Yes. I also considered this. Since another guy supported this, let's do it...

Reply:
Changed :-)

Reply:
Once again, consider rsa->meth that implements only keygen and application calling RSA_generate_multi_prime_key. I argue that it should return error, as opposite to using rsa_builtin_keygen. Because you can't expect rsa->meth that doesn't implement multi_prime_keygen to know what to do with key generated by builtin generator.

Reply:
> Because you can't expect rsa->meth that doesn't implement multi_prime_keygen to know what to do with key generated by builtin generator.

Seems this is the original behaviour of RSA_generate_key_ex.

Reply:
> this is original behaviour of RSA_generate_key_ex

Yes, and it was restored by chaining call to RSA_generate_multi_prime_key (with primes == 2). But question is not about RSA_generate_key_ex, but *application* calling *new* RSA_generate_multi_prime_key (naturally with primes > 2 when there is no original behaviour to preserve) and ending up with rsa->meth provider that doesn't implement multi_prime_keygen. What could rsa->meth possibly mean by that? If rsa->meth->rsa_keygen method is not NULL, then it's safe to assume that rsa->meth expects it to be called in two-prime case. Even when application calls RSA_generate_multi_prime_key with primes == 2. And [once again] question is if it would be appropriate to call builtin key generator for primes > 2 in such case? I.e. when we see that rsa->meth is explicitly two-prime? Note that  builtin generator will set up p and q, so that two-prime rsa->meth can perform calculations on them. Calculations will fail to pass verification with n, but do we actually want to give an impression that rsa->meth could attempt to perform them?